pkg/apis/validation/validation.go

```diff
@@ -113,11 +115,59 @@ func boolValue(path *field.Path, s string) field.ErrorList {
 	return nil
 }
 
-func keyEncodingValue(path *field.Path, s string) field.ErrorList {
-	if s != string(cmapi.PKCS1) && s != string(cmapi.PKCS8) {
-		return field.ErrorList{field.NotSupported(path, s, []string{string(cmapi.PKCS1), string(cmapi.PKCS8)})}
+func keyValue(path *field.Path, attr map[string]string) field.ErrorList {
+	algPath := path.Child(csiapi.KeyAlgorithmKey)
+	encodingPath := path.Child(csiapi.KeyEncodingKey)
+	sizePath := path.Child(csiapi.KeySizeKey)
+
+	alg := attr[csiapi.KeyAlgorithmKey]
+	encoding := attr[csiapi.KeyEncodingKey]
+	size := attr[csiapi.KeySizeKey]
+
+	if encoding != string(cmapi.PKCS1) && encoding != string(cmapi.PKCS8) {
+		return field.ErrorList{field.NotSupported(encodingPath, encoding, []string{string(cmapi.PKCS1), string(cmapi.PKCS8)})}
+	}
+
+	switch alg {
+	case string(cmapi.RSAKeyAlgorithm):
+		s, err := strconv.Atoi(size)
+		if err != nil {
+			return field.ErrorList{field.Invalid(sizePath, size, err.Error())}
+		}
+		if s < pki.MinRSAKeySize {
+			return field.ErrorList{field.Invalid(sizePath, size, fmt.Sprintf("key size must be at least %d when using the RSA key algorithm", pki.MinRSAKeySize))}
+		}
+		if s > pki.MaxRSAKeySize {
+			return field.ErrorList{field.Invalid(sizePath, size, fmt.Sprintf("key size must not exceed %d when using the RSA key algorithm", pki.MaxRSAKeySize))}
+		}
+		return nil
+	case string(cmapi.ECDSAKeyAlgorithm):
+		if encoding == string(cmapi.PKCS1) {
+			return field.ErrorList{field.Invalid(encodingPath, encoding, "pkcs1 cannot be used with ecdsa. use pkcs8 or pkcs12 instead")}
+		}
+		s, err := strconv.Atoi(size)
+		if err != nil {
+			return field.ErrorList{field.Invalid(sizePath, size, err.Error())}
+		}
+		switch s {
+		case pki.ECCurve256:
+		case pki.ECCurve384:
+		case pki.ECCurve521:
+		default:
+			return field.ErrorList{field.NotSupported(sizePath, size, []string{strconv.Itoa(pki.ECCurve256), strconv.Itoa(pki.ECCurve384), strconv.Itoa(pki.ECCurve521)})}
+		}
+		return nil
+	case string(cmapi.Ed25519KeyAlgorithm):
```

Comment:
warning: This string is "Ed25519" (multiple cased letters) so if we use ToUpper on `alg` we'd also need to use ToUpper on `cmapi.Ed25519KeyAlgorithm` also or the comparison will never be true!